libbeat/autodiscover/providers/jolokia/discovery.go

```diff
@@ -0,0 +1,269 @@
+	}
+	return event
+}
+
+type Instance struct {
+	LastSeen time.Time
+	Message  common.MapStr
+}
+
+type Discovery struct {
+	sync.Mutex
+
+	Interfaces   []string
+	Period       time.Duration
+	ProbeTimeout time.Duration
+	GracePeriod  time.Duration
+
+	instances map[string]*Instance
+
+	events chan Event
+	stop   chan struct{}
+}
+
+func (d *Discovery) Start() {
+	d.instances = make(map[string]*Instance)
+	d.events = make(chan Event)
+	d.stop = make(chan struct{})
+	go d.run()
+}
+
+func (d *Discovery) Stop() {
+	d.stop <- struct{}{}
+	close(d.events)
+}
+
+func (d *Discovery) Events() chan Event {
+	return d.events
+}
+
+func (d *Discovery) run() {
+	for {
+		d.sendProbe()
+		d.checkStopped()
+
+		select {
+		case <-time.After(d.Period):
+		case <-d.stop:
+			return
+		}
+	}
+}
+
+// TODO: Check if this can be reused, or if packetbeat has something for this
+func (d *Discovery) interfaces() ([]net.Interface, error) {
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+	matching := make(map[string]net.Interface)
+	for _, name := range d.Interfaces {
+		if name == "any" {
+			return interfaces, nil
+		}
+
+		for _, i := range interfaces {
+			if _, found := matching[i.Name]; !found && matchInterfaceName(name, i.Name) {
+				matching[i.Name] = i
+			}
+		}
+	}
+
+	r := make([]net.Interface, 0, len(matching))
+	for _, i := range matching {
+		r = append(r, i)
+	}
+	return r, nil
+}
+
+func matchInterfaceName(name, candidate string) bool {
+	if strings.HasSuffix(name, "*") {
+		return strings.HasPrefix(candidate, strings.TrimRight(name, "*"))
+	}
+	return name == candidate
+}
+
+func getIPv4Addr(i net.Interface) (net.IP, error) {
+	addrs, err := i.Addrs()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get addresses for "+i.Name)
+	}
+	for _, a := range addrs {
+		if ip, _, err := net.ParseCIDR(a.String()); err == nil && ip != nil {
+			if ipv4 := ip.To4(); ipv4 != nil {
+				return ipv4, nil
+			}
+		}
+	}
+	return nil, nil
+}
+
+var discoveryAddress = net.UDPAddr{IP: net.IPv4(239, 192, 48, 84), Port: 24884}
+var queryMessage = []byte(`{"type":"query"}`)
+
+func (d *Discovery) sendProbe() {
+	interfaces, err := d.interfaces()
+	if err != nil {
+		logp.Err("failed to get interfaces: ", err)
+		return
+	}
+
+	var wg sync.WaitGroup
+	for _, i := range interfaces {
+		ip, err := getIPv4Addr(i)
+		if err != nil {
+			logp.Err(err.Error())
+			continue
+		}
+		if ip == nil {
+			continue
+		}
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			conn, err := net.ListenPacket("udp4", net.JoinHostPort(ip.String(), "0"))
+			if err != nil {
+				logp.Err(err.Error())
+				return
+			}
+			defer conn.Close()
+			conn.SetDeadline(time.Now().Add(d.ProbeTimeout))
+
+			if _, err := conn.WriteTo(queryMessage, &discoveryAddress); err != nil {
+				logp.Err(err.Error())
+				return
+			}
+
+			b := make([]byte, 1500)
+			for {
+				n, _, err := conn.ReadFrom(b)
+				if err != nil {
+					if !err.(net.Error).Timeout() {
```

Comment:
probably you need a type check/switch here

Reply:
In principle `err` here is always going to be `nil` or of type `net.Error`, but yes, just in case it can be a good idea to properly check the type :)